src/VaultFed.sol

```diff
@@ -0,0 +1,140 @@
+    function mint(address to, uint amount) external;
+    function burn(uint amount) external;
+    function balanceOf(address owner) external view returns (uint256);
+    function transfer(address to, uint256 value) external returns (bool);
+    function transferFrom(address from, address to, uint256 value) external returns (bool);
+}
+
+contract VaultFed {
+
+    address public gov;
+    address public pendingGov;
+    address public chair;
+    IERC4626 public immutable vault;
+    IERC20 public immutable dola;
+
+    uint public supply;
+    uint public supplyCap;
+
+    constructor(address _vault, address _gov, address _chair) {
+        require(_vault != address(0), "Vault is required");
+        require(_gov != address(0), "Gov is required");
+        require(_chair != address(0), "Chair is required");
+        require(block.chainid == 1, "Must mint DOLA on Mainnet");
+        vault = IERC4626(_vault);
+        gov = _gov;
+        chair = _chair;
+        dola = IERC20(vault.asset());
+        dola.approve(address(vault), type(uint).max);
+    }
+
+    modifier onlyGov() {
+        require(msg.sender == gov, "Only gov can call this");
+        _;
+    }
+
+    modifier onlyChair() {
+        require(msg.sender == chair || msg.sender == gov, "NOT PERMISSIONED");
+        _;
+    }
+
+    function setPendingGov(address _pendingGov) onlyGov external {
+        pendingGov = _pendingGov;
+        emit NewPendingGov(_pendingGov);
+    }
+
+    function acceptGov() external {
+        require(msg.sender == pendingGov, "NOT PENDING GOV");
+        gov = pendingGov;
+        pendingGov = address(0);
+        emit NewGov(gov);
+    }
+
+    function setChair(address newChair) onlyGov external{
+        chair = newChair;
+        emit NewChair(newChair);
+    }
+
+    function setSupplyCap(uint newSupplyCap) onlyGov external {
+        supplyCap = newSupplyCap;
+        emit NewSupplyCap(newSupplyCap);
+    }
+
+    function resign() onlyChair external {
+        chair = address(0);
+        emit NewChair(address(0));
+    }
+
+    function expansion(uint amount) onlyChair external {
+        require(supply + amount <= supplyCap, "Supply cap exceeded");
+        supply += amount;
+        dola.mint(address(this), amount);
+        vault.deposit(amount, address(this));
+        emit Expansion(amount);
+    }
+
+    function contraction(uint amount) onlyChair external {
+        supply -= amount; // reverts when burning profits
+        vault.withdraw(amount, address(this), address(this));
+        dola.burn(amount);
+        emit Contraction(amount);
+    }
+
+    function contractAll() onlyChair external {
+        uint shares = vault.convertToShares(supply);
+        uint bal = vault.balanceOf(address(this));
+        if(shares > bal) {
+            shares = bal;
+        }
+        uint assets = vault.redeem(shares, address(this), address(this));
+        supply -= assets;
+        dola.burn(assets);
+        emit Contraction(assets);
+    }
```

Comment:
Considering `contractAll` is mostly intended for emergency situations or winding down the fed, wouldn't it be better to exit all shares if possible, then repay as much of the supply as possible and take the rest as profit?
```suggestion
    function contractAll() onlyChair external {
        uint shares = vault.balanceOf(address(this));
        uint assets = vault.redeem(shares, address(this), address(this));
        if(assets <= supply){
            supply -= assets;
            dola.burn(assets);
            emit Contract(assets);
        } else {
            assets -= supply;
            dola.burn(supply);
            emit Contraction(supply);
            supply = 0;
            dola.transfer(gov, assets);
            emit Profit(assets);
        }
    }
```

Reply:
I've removed contractAll entirely as it is redundant with contraction and introduces unnecessary complexity.